### PR TITLE
Support error types

### PR DESCRIPTION
**Pull request by akutuev**
This is an enhancement related to that discussion: #238 

The purpose of this PR is to make a possibility convert an error to its specific type.

Since the errors aggregation has been added before (#233), user should convert an error result to AggregateError first and then iterate over accumulated errors, example:

```
        err := Parse(&config{})
	if e, ok := err.(*AggregateError); ok {
		for _, er := range e.Errors {
			switch v := er.(type) {
			case ParseError:
				// handle it
			case NotStructPtrError:
				// handle it
			case NoParserError:
				// handle it
			case NoSupportedTagOptionError:
				// handle it
			default:
				fmt.Printf("Unknown type %v", v)
			}
		}
	}
```

### CODE REVIEW

**Comment by akutuev** [comment]
@caarlos0 @jurajkulich please feel free to review and validate it

Any comments appreciated 

Thanks

**Review by caarlos0 — COMMENTED** [review]
a few comments, but looking good so far! 
thanks for the PR :D

File: error.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	msg string
+}
+
+func newNoParserError(sf reflect.StructField) error {
+	return NoParserError{fmt.Sprintf(`no parser found for field "%s" of type "%s"`, sf.Name, sf.Type)}
+}
+
+func (e NoParserError) Error() string {
+	return e.msg
+}
```
[the remark this answers is not in the document]
Comment (akutuev) [reply error.go:78]: fair comment, I wanted to make all errors generic: the same structure (msg only), but your comment makes sense:
- User converts error to specific type of the error
- User now has access to all error fields +message string

Let me improve it

File: error.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	msg string
+}
+
+func newNotStructPtrError() error {
+	return NotStructPtrError{"expected a pointer to a Struct"}
+}
+
+func (e NotStructPtrError) Error() string {
+	return e.msg
+}
```
[the remark this answers is not in the document]
Comment (akutuev) [reply error.go:66]: that's definitely true, but I did it deliberately to follow the same error pattern like other errors (struct + func to create and error) do you recommend to make it simpler?

[the remark this answers is not in the document]
Comment (caarlos0) [reply error.go:66]: yes, simpler is usually better :)

**Comment by akutuev** [comment]
I'm sorry, I'm on the holiday till 04.01.23, will review and update PR when I back

**Comment by akutuev** [comment]
Hello @caarlos0. Can you please take a look at the changes when you have time

File: error.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (akutuev) [reply error.go]: I came to realize we are loosing a possibility to check the error type using a errors.New(). But anyway, I have updated this error, can you please check

File: error.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (akutuev) [reply error.go]: Updated as discussed

**Comment by caarlos0** [comment]
sorry, got caught up in some personal issues, will review it when I get some peace time

**Pull request merged**